fw/http_msg.c

```diff
@@ -815,6 +820,120 @@ __hdr_sub(TfwHttpMsg *hm, const TfwStr *hdr, unsigned int hid)
 	return TFW_PASS;
 }
 
+typedef struct {
+	struct sk_buff *skb;
+	const char *data;
+	unsigned int len;
+	unsigned int tail;
+	char *data_off;
+} replace_string_t;
+
+/**
+ * Callback for processing HTTP method substitution
+ */
+static int
+tfw_http_req_meth_sub(void *data, unsigned char *buf, size_t len,
+		  unsigned int *read)
+{
+	int r = T_OK;
+	replace_string_t *replace = data;
+
+	unsigned int off = 0;
+
+	BUG_ON(!buf);
+	BUG_ON(!len);
+	BUG_ON(!replace);
+	BUG_ON(!replace->data);
+
+	if (replace->len > 0) {
+		off = min_t(unsigned int, len, replace->len);
+		memcpy_fast(buf, replace->data, (size_t)off);
+		*read += off;
+
+		if (len >= replace->len) {
+			replace->data_off = buf+off;
+			replace->len = 0;
+			r = T_OK;
+		} else {
+			replace->len -= len;
+			r = T_POSTPONE;
+		}
+	}
+	if (!len) {
+		r = T_POSTPONE;
+	}
+
+	return r;
+}
+
+/**
+ * Substitute request method with "get"
+ *
+ * Note: replaced method name length is supposed to be shorter
+ *       than the replacement string ("get")
+ */
+int
+tfw_http_req_meth_sub_with_get(TfwHttpReq *req)
+{
+	const char *dest_name;
+	unsigned int dest_len;
+
+	const char *sub_name = "GET";
+	unsigned int sub_len = sizeof("GET")-1;
+
+	struct sk_buff *skb;
+	unsigned int tail;
+
+	unsigned int parsed, unused;
+	replace_string_t replace;
+
+	int r;
+
+	TfwStr it = {};
+	int _;
+
+	BUG_ON(!req);
+
+	dest_name = tfw_http_tbl_method_get_name(req->method);
+	BUG_ON(!dest_name);
+
+	dest_len = strlen(dest_name);
+
+	BUG_ON(dest_len < sub_len);
+
+	tail = dest_len - sub_len;
+
+	BUG_ON(!req->msg.skb_head);
+
+	skb = req->msg.skb_head;
```

Comment:
I rewrote the function to get deeper with skb's and chunks.
and renamed it to tfw_http_subst_purge_with_get to describe it better. 
for a while it's designed for one task only, but it can be turned into something generic later. 
I work on tests for it.